src/graphic/text/rt_render.cc

```diff
@@ -1322,15 +1350,100 @@ class PTagHandler : public TagHandler {
 	uint16_t indent_;
 };
 
+class LinkTagHandler : public TagHandler {
+public:
+	LinkTagHandler(const TagHandler* p,
+	               Tag& tag,
+	               FontCache& fc,
+	               const NodeStyle& ns,
+	               ImageCache* image_cache,
+	               RendererStyle& init_renderer_style,
+	               const UI::FontSets* fontsets)
+	   : TagHandler(p, tag, fc, ns, image_cache, init_renderer_style, fontsets) {
+	}
+
+	enum class Type { kURL, kUI };
+
+	void enter() override {
+		const AttrMap& a = tag_.attrs();
+
+		std::string t = a["type"].get_string();
+		if (t == "url") {
+			type_ = Type::kURL;
+		} else if (t == "ui") {
+			type_ = Type::kUI;
+			action_ = a["action"].get_string();
+		} else {
+			throw RenderError(format("Invalid link type '%s'", t));
+		}
+		target_ = a["target"].get_string();
+
+		if (a.has("mouseover")) {
+			mouseover_ = a["mouseover"].get_string();
+		}
+
+		if (a.has("align")) {
+			const std::string align = a["align"].get_string();
+			if (align == "right") {
+				nodestyle_.halign = UI::Align::kRight;
+			} else if (align == "center" || align == "middle") {
+				nodestyle_.halign = UI::Align::kCenter;
+			} else {
+				nodestyle_.halign = UI::Align::kLeft;
+			}
+		}
+		nodestyle_.halign = mirror_alignment(nodestyle_.halign, nodestyle_.is_rtl);
+		if (a.has("valign")) {
+			const std::string align = a["valign"].get_string();
+			if (align == "bottom") {
+				nodestyle_.valign = UI::Align::kBottom;
+			} else if (align == "center" || align == "middle") {
+				nodestyle_.valign = UI::Align::kCenter;
+			} else {
+				nodestyle_.valign = UI::Align::kTop;
+			}
+		}
+		if (a.has("spacing")) {
+			nodestyle_.spacing = a["spacing"].get_int(std::numeric_limits<uint8_t>::max());
+		}
+	}
+
+	bool handle_mousepress(int32_t /* x */, int32_t /* y */) const override {
+		switch (type_) {
+		case Type::kUI:
+			Notifications::publish(NoteHyperlink(target_, action_));
+			return true;
+		case Type::kURL:
+#if SDL_VERSION_ATLEAST(2, 0, 14)
+			SDL_OpenURL(target_.c_str());
+#else
+			SDL_SetClipboardText(target_.c_str());
+#endif
```

Comment:
Can we open a notification window from here to tell the user when the link was copied to the clipboard instead of opening it? Or actually use `g_can_open_url` both here and in `as_url_hyperlink(url)` to change the mouseover text to "Copy to clipboard"? Maybe the second one is nicer to the user and not worse in code either.